src/AndersonAcceleration.f90

```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
+        CALL BLAS_axpy(solver%tmpvec,x_new,solver%alpha(i))
+      ENDDO
+
+      !Push back solution vectors and load in newest accelerated as next initial iterate
+      DO i=MIN(depth_s+1,solver%depth),1,-1
+        CALL BLAS_copy(solver%x(i),solver%x(i+1))
+      ENDDO
+      CALL BLAS_copy(x_new,solver%x(1))
+    ENDIF
   ELSE
-    CALL Anderson_Update(solver%id)
+    !Get under-relaxed solution using mixing parameter
+    CALL BLAS_scal(x_new,solver%beta)
+    CALL BLAS_axpy(solver%x(1),x_new,1.0_SRK-solver%beta)
+    CALL BLAS_copy(x_new,solver%x(1))
   ENDIF
 
-  solver%iter=solver%iter+1
-#endif
 ENDSUBROUTINE step_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Reset the Anderson acceleration solver
-!> @param solver The anderson solver to act on
-!>
-!> This routine resets the counters for the anderson acceleration
+!> @brief Set or reset the initial iterate and linear solver state for the Anderson solver
+!> @param solver The Anderson solver to act on
+!> @param x Initial iterate solve is starting from
 !>
-SUBROUTINE reset_AndersonAccelerationType(solver)
+SUBROUTINE reset_AndersonAccelerationType(solver,x)
+  CHARACTER(LEN=*),PARAMETER :: myName='reset_AndersonAccelerationType'
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  CALL Anderson_Reset(solver%id)
-  solver%iter=1
+  CLASS(VectorType),INTENT(INOUT) :: x
+
+  INTEGER(SIK) :: i,j
+
+  REQUIRE(x%n == solver%n)
+
+  !If this is the first call to set/reset must actually create vectors of needed type
+  IF(solver%s == 0) THEN
+    IF(.NOT.ALLOCATED(solver%x)) THEN
+
+      !!!TODO Once missing BLAS interfaces have been added for the following vector types
+      !do away with this error catch to allow use of these with Anderson Acceleration.
+#ifdef HAVE_MPI
+      SELECT TYPE(x);TYPE IS(NativeDistributedVectorType)
+        CALL solver%ce%exceptHandler%raiseError('Incorrect call to '//modName// &
+            '::'//myName//' - Input vector type not supported!')
 #endif
+#ifdef FUTILITY_HAVE_Trilinos
+      TYPE IS(TrilinosVectorType)
+        CALL solver%ce%exceptHandler%raiseError('Incorrect call to '//modName// &
+            '::'//myName//' - Input vector type not supported!')
+      ENDSELECT
+#endif
+
+      CALL VectorResembleAlloc(solver%x,x,solver%depth+1)
+      CALL VectorResembleAlloc(solver%Gx,x,solver%depth+1)
+      CALL VectorResembleAlloc(solver%r,x,solver%depth+1)
+      CALL VectorResembleAlloc(solver%tmpvec,x)
+    ELSE
+      CALL solver%ce%exceptHandler%raiseError('Incorrect call to '//modName// &
+          '::'//myName//' - At least one step required before reseting!')
+    ENDIF
+  ENDIF
+
+  !Reset iteration counter
+  solver%s=0_SIK
+
+  !Grab initial iterate to initiate Anderson from
+  CALL BLAS_copy(x,solver%x(1))
+
+  !Reset Anderson coefficient matrix
+  IF(solver%LS%isinit) THEN
```

Comment:
Should we just REQUIRE this?  Seems like you shouldn't be calling reset on something that was never initialized to begin with

Reply:
No, not for this guy. If depth <= 1 the linear solver is never inited because it isn't needed since the alphas can be calculated with a simple formula.

Reply:
Oh, gotcha.  I missed the `%LS` part.  Thought that was the Anderson object you were checking.